src/mysql-util/src/decoding.rs

```diff
@@ -0,0 +1,188 @@
+                temp_strs.push(val);
+                Datum::from(temp_strs.last().unwrap().as_str())
+            }
+            ScalarType::VarChar { max_length } => {
+                let val = from_value_opt::<String>(value)?;
+                check_char_length(max_length.map(|l| l.into_u32()), &val, col_desc)?;
+                temp_strs.push(val);
+                Datum::from(temp_strs.last().unwrap().as_str())
+            }
+            ScalarType::String => {
+                temp_strs.push(from_value_opt::<String>(value)?);
+                Datum::from(temp_strs.last().unwrap().as_str())
+            }
+            ScalarType::Bytes => {
+                let data = from_value_opt::<Vec<u8>>(value)?;
+                temp_bytes.push(data);
+                Datum::from(temp_bytes.last().unwrap().as_slice())
+            }
+            ScalarType::Date => {
+                let date = Date::try_from(from_value_opt::<chrono::NaiveDate>(value)?)?;
+                Datum::from(date)
+            }
+            ScalarType::Timestamp { precision: _ } => {
+                // Timestamps are encoded as different mysql_common::Value types depending on
+                // whether they are from a binlog event or a query, and depending on which
+                // mysql timestamp version is used. We handle those cases here
+                // https://github.com/blackbeam/rust_mysql_common/blob/v0.31.0/src/binlog/value.rs#L87-L155
+                // https://github.com/blackbeam/rust_mysql_common/blob/v0.31.0/src/value/mod.rs#L332
+                let chrono_timestamp = match value {
+                    Value::Date(..) => from_value_opt::<chrono::NaiveDateTime>(value)?,
+                    // old temporal format from before MySQL 5.6; didn't support fractional seconds
+                    Value::Int(val) => chrono::NaiveDateTime::from_timestamp_opt(val, 0)
+                        .ok_or(anyhow::anyhow!("received invalid timestamp value: {}", val))?,
+                    Value::Bytes(data) => {
+                        let data = std::str::from_utf8(&data)?;
+                        if data.contains('.') {
+                            chrono::NaiveDateTime::parse_from_str(data, "%s%.6f")?
+                        } else {
+                            chrono::NaiveDateTime::parse_from_str(data, "%s")?
+                        }
+                    }
+                    _ => Err(anyhow::anyhow!(
+                        "received unexpected value for timestamp type: {:?}",
+                        value
+                    ))?,
+                };
+                Datum::try_from(CheckedTimestamp::try_from(chrono_timestamp)?)?
+            }
+            ScalarType::Time => Datum::from(from_value_opt::<chrono::NaiveTime>(value)?),
+            ScalarType::Numeric { max_scale } => {
+                let val = from_value_opt::<String>(value)?;
+                let val = Numeric::from_str(&val)?;
```

Comment:
I think we also need to check that `get_precision(&val) <= NUMERIC_DATUM_MAX_PRECISION` here